### PR TITLE
[WIP] Don't reuse connections when the host changed to be compatible with SNI headers

### PR DESCRIPTION
**Pull request by luizhespanha**
Hello, this is still a WIP and only a POC to collect feedback! If you think this code makes sense I'll try to add some tests.

To give the context of the problem:
We are using an SNI proxy, so for example in AWS Route 53 I have:
my-service-a.xpto.com -> my-sni-proxy.xpto.com
my-service-b.xpto.com -> my-sni-proxy.xpto.com

A client makes a request to my-service-a.xpto.com, then my-sni-proxy receives this request, looks on the SNI headers, and sees that the request if for `my-service-a` and dispatch the request.

The problem that we are facing is:
A client makes a request to my-service-a.xpto.com using httpkit and with keepalive enabled. The request is made and the connection is added to keepalives to be reused, but the addr stored it the IP and PORT of my sni-proxy. Then the client makes a request to my my-service-b.xpto.com in the sequence and as it will also use the proxy, the connection will be reused with the same SNI header, which is my-service-a.xpto.com, and not my-service-b.xpto.com, so the proxy sends the request to the wrong destination.

Change the keepalive to -1 solves this problem for us, but it's tricky for the users to understand and sometimes they even know that the SNI proxy exists.

So I changed the code a little bit to consider the host and the IP and PORT together to decide if we should or not reuse a connection. So, in my example, when the client makes a request to my-service-b.xpto.com, the key will be my-service-b.xpto.com123.123.123.123:443 so we won't reuse the same connection used to invoke my-service-a, but it the client tries to invoke the service-a again, I will be able to reuse the connection.

WDYT? Does it make sense? Do you have some other suggestions to deal with it?

Thanks!

### CODE REVIEW

**Comment by ptaoussanis** [comment]
@luizhespanha Hi Luiz! Thanks a lot for the clear context, and PR.

My quick first impression is that your approach here sounds reasonable. Would be up for seeing some tests added 👍 

Cheers :-)

File: src/java/org/httpkit/client/HttpClient.java [hunk 1]
```diff
@@ -309,8 +310,10 @@ public void exec(String url, RequestConfig cfg, SSLEngine engine, IRespListener
         try {
             if (proxyUri == null) {
                 addr = addressFinder.findAddress(uri);
+                host = uri.getHost();
             } else {
                 addr = addressFinder.findAddress(proxyUri);
+                host = proxyUri.getHost();
```
Comment (andrerocker) [inline src/java/org/httpkit/client/HttpClient.java:316]: In this case, I think that we need to use something like `proxyUri.getHost() + uri.getHost()`

**Comment by bsless** [comment]
Hi all, what's the status of this PR? 
We have some interest in it as well. If you'd like a few more eyes to look it over we can dedicate some people to it (we: @Appsflyer)

**Comment by ptaoussanis** [comment]
@bsless Hi Ben!

Not sure if Luiz is still available for or interested in this PR (@luizhespanha?) - but next step would be for someone to review and add tests. (Which would be very welcome!)

Cheers :-)

**Review by Queen300 — APPROVED** [review]
[H](d)

**Comment by maorelias** [comment]
Hi guys any updates on this? 
I'm also suffering from this issue and I think it's important to fix it because it a counter-intuitive behavior - created the http client with the `sni-client/ssl-configurer` but since keep-alive is enabled, the SNI headers get mixed up from time to time when sending requests to multi-tenant IPs.

**Comment by ptaoussanis** [comment]
@luizhespanha Hi Luiz, it would be great to try and get this work included as part of an upcoming v2.7 release. Do you still have any interest in this? Is there something I or someone else might be able to do to try and assist?

**Comment by ptaoussanis** [comment]
Last call for assistance on this PR, in the hope of possible inclusion for upcoming v2.7.

Things that could help get this merged (in order of low->high effort):

- 👍 that the approach in this PR looks solid (along with @andrerocker's [suggestion](https://github.com/http-kit/http-kit/pull/446#discussion_r488711750)).
- Confirmation if you've successfully used these changes yourself.
- Added unit tests.

Thanks everyone!

**Comment by ptaoussanis** [comment]
PR merged so long for experimental inclusion in v2.7 beta.
Closing for now - but feel free to re-open or add comments if anyone still has any input.

@luizhespanha Thanks for all your work Luiz! 🚀

**Pull request closed without merging**